File: smart-agriculture/backend/app/services/auth_service.py
from __future__ import annotations

from datetime import datetime

from fastapi import HTTPException, status
from sqlalchemy.orm import Session

from app.core.security import create_access_token, create_refresh_token, get_password_hash, verify_password
from app.models.user import User
from app.schemas.auth import RegisterRequest


class AuthService:
    def __init__(self, db: Session):
        self.db = db

    def register(self, data: RegisterRequest) -> User:
        existing = self.db.query(User).filter(User.email == data.email.lower()).first()
        if existing:
            raise HTTPException(status_code=status.HTTP_409_CONFLICT, detail="User already exists")

        user = User(
            full_name=data.full_name,
            email=data.email.lower(),
            phone=data.phone,
            password_hash=get_password_hash(data.password),
            role=data.role.lower() if data.role else "farmer",
            location=data.location,
            is_active=True,
        )
        self.db.add(user)
        self.db.commit()
        self.db.refresh(user)
        return user

    def login(self, email: str, password: str):
        user = self.db.query(User).filter(User.email == email.lower(), User.is_active.is_(True)).first()
        if not user or not verify_password(password, user.password_hash):
            raise HTTPException(status_code=status.HTTP_401_UNAUTHORIZED, detail="Invalid email or password")

        access_token = create_access_token(str(user.id), user.role)
        refresh_token = create_refresh_token(str(user.id), user.role)
        return {"user": user, "access_token": access_token, "refresh_token": refresh_token}

    def get_profile(self, user_id: int) -> User:
        user = self.db.query(User).filter(User.id == user_id).first()
        if not user:
            raise HTTPException(status_code=status.HTTP_404_NOT_FOUND, detail="User not found")
        return user

File: smart-agriculture/frontend/src/services/authService.ts
import { supabase } from '../lib/supabase'

export const authService = {
  async signUp(payload: {
    email: string
    password: string
    full_name: string
    phone?: string
    location?: string
  }) {
    const { data, error } = await supabase.auth.signUp(
      {
        email: payload.email,
        password: payload.password,
      },
      { data: { full_name: payload.full_name, phone: payload.phone, location: payload.location } },
    )
    return { data, error }
  },
  async signIn(payload: { email: string; password: string }) {
    const { data, error } = await supabase.auth.signInWithPassword({ email: payload.email, password: payload.password })
    return { data, error }
  },
  async signOut() {
    const { error } = await supabase.auth.signOut()
    return { error }
  },
  async getUser() {
    const { data, error } = await supabase.auth.getUser()
    return { data, error }
  },
}
